ide/coqide/idetop.ml

```diff
@@ -367,6 +367,17 @@ let export_option_state s = {
   Interface.opt_value = export_option_value s.Goptions.opt_value;
 }
 
+exception NotSupported of string
+
+let proof_diff (diff_opt, sid) =
+  let diff_opt = Proof_diffs.string_to_diffs diff_opt in
+  let doc = Stm.get_doc 0 in
+  match Stm.get_proof ~doc sid with
+  | None -> Pp.str "No proofs to diff"
+  | Some proof ->
```

Comment:
Actually would be more sensible to have this method return an `option` type directly.

Reply:
Huh?  This is the protocol handler for this type of message, referenced only by `handler`, which is used by `Xmlprotocol.abstract_eval_call`.  In that routine, every message type has a single line in the form `| PDiff x      -> mkGood (handler.proof_diff x)`.  Seems to me you wouldn't want to break the symmetry by adding message-specific logic there.

Reply:
Actually this removes flexibility from the IDE, if you return in the protocol an option type the IDE can indeed handle the case better, otherwise it has little choice, as for example to update / translate / beautify the diff message.

Reply:
I don't see how this removes flexibility.  The code is easily refactored when there is a need to update/translate/beautify the diff message.  It's work for the future that's not needed now.  Even so, why couldn't the update/translate/beautify go into this routine?

@gares, what do you suggest?

Reply:
I think the point here is that if the XML protocol message carries an option pp.t then the UI can tell if there is a diff or not, and say, open a popup or not. By sending always a pp.t the UI cannot easily tell if it is a real diff or not, and react accordingly. IIRC we have support for the option type in the XML encoder, so it should be doable.

Reply:
At some extent, sending back the empty string makes it easier to detect. And None is an idiomatic way of representing it.

Reply:
Yeah what @gares said, I think making this into an option is cleaner, for example such calls in jsCoq / SerAPI protocol all of them return option so my experience in interfaces tells me that's the right thing to do.

Reply:
Frequently, showing a popup is overkill.  It's distracting.  It's a bit dated.  For example, if a command such as this one normally puts output in the Messages panel, it's better to display its error message in the same place--which is where the user will be looking anyway.  No need for a popup.

Some/None doesn't work if there's more than one error message that can be returned.

I think the text of the error message should come from the server so it's the same message across different IDEs.  Returning None puts the burden of discovering what None really means and devising wording on each IDE developer, which seems misplaced.

What do you think?